react-basic-practice/src/features/SignUpForm/index.jsx

```diff
@@ -1,21 +1,84 @@
-import React, { useRef, useState } from 'react';
+import React, { useState } from 'react';
+import { Link } from 'react-router-dom';
 import { v4 as uuidv4 } from 'uuid';
 
+import { REGEXP } from '../../constants/regexp';
+
 import Button from '../../components/common/Button';
 import Input from '../../components/common/Input';
 import Logo from '../../components/common/Logo';
 import FormGroup from '../../components/FormGroup';
 import './index.css';
 
 const SignUpForm = () => {
+  const initialErrorMsgs = {
+    inputEmail: '',
+    inputUserName: '',
+    inputPassword: '',
+  };
+  const initialInput = {
+    inputEmail: '',
+    inputUserName: '',
+    inputPassword: '',
+  };
   const [isSignUpLoading, setIsSignUpLoading] = useState(false);
-  // const [errorMessage, setErrorMessage] = useState({});
-  const inputUsername = useRef(null);
-  const inputPassword = useRef(null);
-  const inputEmail = useRef(null);
+  const [errorMessage, setErrorMessage] = useState(initialErrorMsgs);
+  const [inputValue, setInputValue] = useState(initialInput);
+
+  const handleInputValue = (value) => {
+    setInputValue({ ...inputValue, ...value });
+  };
+
+  const checkEmpty = () => {
+    for (const key in inputValue) {
+      if (inputValue[key] === '') {
+        setErrorMessage((preMsg) => ({
+          ...preMsg,
+          [key]: 'This input is required.',
+        }));
+      } else {
+        setErrorMessage((preMsg) => ({
+          ...preMsg,
+          [key]: '',
+        }));
+      }
+    }
+  };
+
+  const checkValidate = () => {
+    if (inputValue.inputPassword !== '') {
+      REGEXP.REGEXP_PASSWORD.test(inputValue.inputPassword)
+        ? setErrorMessage((preMsg) => ({ ...preMsg, inputPassword: '' }))
+        : setErrorMessage((preMsg) => ({ ...preMsg, inputPassword: 'Invalid password' }));
+    }
 
-  const handleSignUp = () => {
+    if (inputValue.inputEmail !== '') {
+      REGEXP.REGEXP_MAIL.test(inputValue.inputEmail)
+        ? setErrorMessage((preMsg) => ({ ...preMsg, inputEmail: '' }))
+        : setErrorMessage((preMsg) => ({ ...preMsg, inputEmail: 'Invalid email' }));
+    }
 
+    if (inputValue.inputUserName !== '') {
+      REGEXP.REGEXP_USER_NAME.test(inputValue.inputUserName)
+        ? setErrorMessage((preMsg) => ({ ...preMsg, inputUserName: '' }))
+        : setErrorMessage((preMsg) => ({ ...preMsg, inputUserName: 'Invalid user name' }));
+    }
+  };
+
+  const handleSignUp = (e) => {
+    try {
+      e.preventDefault();
+
+      setIsSignUpLoading(true);
+
+      checkEmpty();
```

Comment:
checkEmpty is also a form of validation. Should call this in validation function.